Add initial regex source generator

stephentoub opened the pull request:
This adds the `[RegexGenerator(...)]` attribute and a new System.Text.RegularExpressions.Generator source generator for Regex.  Given a method like:
```C#
[RegexGenerator("abc")]
private static partial Regex Abc();
```
it will generate C# code to implement `Abc()` that returns a `Regex` instance with code like you'd have gotten with `RegexOptions.Compiled`, but generated at build time in C# rather than at run time with reflection emit in MSIL.  The generator is largely a port of RegexCompiler.cs; ideally we wouldn't duplicate all of the emit logic, once for MSIL and once for C#, and we should subsequently look for ways to avoid this duplication.  The generator includes source from System.Text.RegularExpressions, in particular in support of parsing and analyzing regexes.

![image](https://user-images.githubusercontent.com/2642209/133515550-1d1bc397-192c-41c3-83c0-eb7479aae400.png)

It includes a test project for the generator focused on testing the mechanics of the generator as well as integrating the source generator into the regex test suite, rolled out into many but not all of the tests so far.

There are a bunch of follow-ups to this for while I'll be opening issues, including rolling this out through the rest of the tests (and finding ways to make it go faster, as it's currently significantly increasing the time for the regex test suite), fixing a bunch of TODOs (e.g. figuring out what to do about culture), and improving perf further (it's now really easy to see opportunities for improved code generation).

Fixes https://github.com/dotnet/runtime/issues/58880
Contributes to https://github.com/dotnet/runtime/issues/44676

cc: @danmoseley, @pgovind, @tannergooding, @eerhardt, @veanes, @olsaarik

stephentoub committed:
Add RegexGeneratorAttribute

Adds the new RegexGenerator attribute that's a signal to the regex
generator to generate code for the specified regex.

stephentoub committed:
Implement RegexGenerator

Add a source generator for generating C# code for Regex.  This is
primarily a port of RegexCompiler.cs, generating C# code instead of
MSIL.

stephentoub committed:
Add generator test project

Adds tests dedicated to the mechanics of the source generator, e.g. that
appropriate diagnostics are issued for improper use of RegexGenerator.

stephentoub committed:
Integrate source generator into many regex tests

Start integrating the source generator into the regex test suite, so
that many existing tests also validate the generated code.

stephentoub commented:
@jaredpar, FYI.

danmoseley commented on src/libraries/System.Text.RegularExpressions/tests/System.Text.RegularExpressions.Generators.Tests/RegexGeneratorParserTests.cs:
```diff
@@ -0,0 +1,432 @@
+        }
+
+        [Fact]
+        public async Task Valid_MultiplRegexMethodsPerClass()
```
```suggestion
        public async Task Valid_MultipleRegexMethodsPerClass()
```
As a temporary optimization, would it make sense to include all the "positive case" regex generator tests above into a single file/single test? (this one basically).. would that speed up the tests noticeably?

danmoseley replied:
and maybe combine some/all of the failing ones, assuming the compiler will return all the diagnostics and we can check the line numbers match expectations.

danmoseley replied:
Never mind, I see we're running the generator on far more patterns. Still I wonder whether combining might be the path to more acceptable perf.

stephentoub replied:
> would that speed up the tests noticeably?

No.  These tests are very fast.  What's adding all the time is we're now running literally thousands of regexes as part of the main test suite, each of which is running the generator, compiling an assembly, etc.

danmoseley commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCode.cs:
```diff
@@ -223,8 +223,7 @@ public static int OpcodeSize(int opcode)
             }
         }
 
-#if DEBUG
```
why remove #if DEBUG?

stephentoub replied:
Because the code is now used as part of the generator in release.  I could change the condition to be something like `#if DEBUG || REGEXGENERATOR`, but the linker will trim it out anyway if nothing uses it, so it didn't seem worth it.

danmoseley commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs:
```diff
@@ -1516,7 +1561,7 @@ public static string SetDescription(string set)
         /// </summary>
-        [ExcludeFromCodeCoverage(Justification = "Debug only")]
+        [ExcludeFromCodeCoverage]
         public static string CharDescription(char ch)
```
BTW, it would be nice to special case `\n` here so it's not \u0a

danmoseley commented on src/libraries/System.Text.RegularExpressions/gen/Resources/Strings.resx:
```diff
@@ -0,0 +1,318 @@
+<?xml version="1.0" encoding="utf-8"?>
```
can we not reuse the existing resx?

stephentoub replied:
How? i.e. is there a mechanism for including an additional resx file and having them merged, or something like that?

danmoseley replied:
I meant just use the existing one - of course not all the strings apply. Perhaps that would get complex given this one is localized.
No I don't know a way to include one in another.

stephentoub replied:
> I meant just use the existing one

And add the generator specific strings into the existing one?

danmoseley replied:
If that's less wasteful than the duplication. Feel free to ignore.

danmoseley commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -0,0 +1,3129 @@
+            // If we can't support custom generation for this regex, spit out a Regex constructor call.
+            if (!SupportsCustomCodeGeneration(rm))
+            {
+                writer.WriteLine($"new Regex({patternExpression}, {optionsExpression}, {timeoutExpression});");
```
should we have a diagnostic for this? they explicitly asked for the generator using the attribute, it's not that we are opportunistically discovering them.

stephentoub replied:
We can decide whether to fail, having a warning diagnostic but still proceed, or just fall back to this.  This won't actually be used right now, as we support all options that make it to this point; I added this so that we don't just fail when NonBacktracking gets added in, just as we don't want RegexOptions.Compiled | RegexOptions.NonBacktracking to start failing.

danmoseley commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -0,0 +1,3129 @@
+                "Timeout.InfiniteTimeSpan" :
+                $"TimeSpan.FromMilliseconds({rm.MatchTimeout.Value.ToString(CultureInfo.InvariantCulture)})";
+
+            writer.WriteLine(s_generatedCodeAttribute);
```
nit, it might be easier to read if you wrapped writer.WriteLine eg `wr(s_generatedCodeAttribute)`

in fact where there's a block of lines written together like here, perhaps it would be clearer to have one giant string so there's just one call to write.

danmoseley commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -0,0 +1,3129 @@
+                            }
+                            else
+                            {
+                                // TODO: This differs subtely between interpreted and compiled. Why?
```
```suggestion
                                // TODO: This differs subtly between interpreted and compiled. Why?
```

danmoseley commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -0,0 +1,3129 @@
+                            break;
+                        }
+
+                    case RegexCode.Oneloop:
```
is there some clever way with pattern matching to avoid all this repetition?

stephentoub replied:
Suggestions welcome.

danmoseley commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -0,0 +1,3129 @@
+
+
+            /// <summary>
+            /// Branch to the MSIL corresponding to the regex code at i
```
MSIL?

stephentoub replied:
It's just a copy/paste of RegexCompiler.  I'll change it.

danmoseley commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -0,0 +1,3129 @@
+
+        private static bool EmitInitializeCultureForGoIfNecessary(IndentedTextWriter writer, RegexMethod rm)
+        {
+            if (((RegexOptions)rm.Options & RegexOptions.CultureInvariant) == 0)
```
tiny nit, I find early return easier to grok generally (and less indented)..

danmoseley replied:
never mind, there's value in mimicking the existing code.. at least for now.b

danmoseley commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -0,0 +1,3129 @@
+            }
+        }
+
+        private static string MatchCharacterClass(bool hasTextInfo, RegexOptions options, string chExpr, string charClass, bool caseInsensitive)
```
a method like this is where it's fairly easy to imagine a single implementation that either emits IL or C# as requested.

stephentoub replied:
Maybe, maybe not.  Even if it is, this is a tiny part of the whole implementation.

I know you care a lot about such sharing.  If you really want to move that forward, I suggest you invest a dev in prototyping it, and using this implementation as the yardstick to measure its success by.

danmoseley replied:
Yup. This code certainly makes that a lot easier: it becomes essentially an exercise in how far we can reasonably reduce code duplication.

danmoseley commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs:
```diff
@@ -0,0 +1,337 @@
+                            regexMethod.Tree = RegexParser.Parse(regexMethod.Pattern, (RegexOptions)regexMethod.Options, culture);
+                            regexMethod.Code = RegexWriter.Write(regexMethod.Tree);
+                        }
+                        catch (Exception e)
```
any exception? why not just RPE and RMTE? anything else is a bug and it would be nice if it was prominent? (eg don't catch all around the regular regex API)

stephentoub replied:
> anything else is a bug and it would be nice if it was prominent? (

It's translating it into a diagnostic.  What would be more prominent?

danmoseley replied:
What happens if we just let the exception escape? I guess I mean - is it best practice for all generators to catch all and log a diagnostic? Whatever is the pattern is fine I guess.

stephentoub replied:
> What happens if we just let the exception escape?

Escape all the way?  The generator simply stops producing code until you restart VS.

danmoseley commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs:
```diff
@@ -0,0 +1,337 @@
+                        }
+                        if (potentialNamespaceParent is BaseNamespaceDeclarationSyntax namespaceParent)
+                        {
+                            nameSpace = namespaceParent.Name.ToString();
```
nit is `nameSpace` the conventional name when using Roslyn API? I'd write `ns` I guess

stephentoub replied:
`ns`, `nameSpace`, `@namespace`, I've seen them all.  No strong preference.

danmoseley commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs:
```diff
@@ -0,0 +1,337 @@
+
+                        bool IsAllowedKind(SyntaxKind kind) =>
+                            kind == SyntaxKind.ClassDeclaration ||
+                            kind == SyntaxKind.StructDeclaration ||
```
tests for struct/record/record struct?

danmoseley commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -1754,7 +1759,170 @@ public int ChildCount()
+        {
+            bool supported = false;
+
+            // We only support the default left-to-right, not right-to-left, which requires more complication in the gerated code.
```
```suggestion
            // We only support the default left-to-right, not right-to-left, which requires more complication in the generated code.
```

danmoseley replied:
of course, this is someone else's typo

danmoseley commented:
Will actually be interesting now we can look at the C# generated for different regexes to more easily see how it works.

stephentoub committed:
Address PR feedback

Szer commented:
Will there be any warning if that library (supposed to be language agnostic?) code would be code NOT from Roslyn?

am11 commented on src/libraries/System.Text.RegularExpressions/tests/Regex.Tests.Common.cs:
```diff
@@ -24,6 +29,65 @@ public static bool IsDefaultStart(string input, RegexOptions options, int start)
+            throw new ArgumentException($"Unknown engine: {engine}");
+        }
+
+        public static Task<Regex> GetRegex(RegexEngine engine, string pattern, RegexOptions options, TimeSpan timeout) =>
```
nit: would it be better to use Async suffix for these methods? `GetRegexAsync`, `SourceGenRegexAsync` etc.

stephentoub replied:
Yup

stephentoub commented:
> Will there be any warning if that library code (supposed to be language agnostic?) would be called NOT from Roslyn?

I don't understand the question.  `Regex` is usable from all .NET languages.  If you try to use `RegexGenerator` with something other than C#, you'll get errors because no code will be generated.  Can you clarify?

Youssef1313 commented on src/libraries/System.Text.RegularExpressions/gen/DiagnosticDescriptors.cs:
```diff
@@ -0,0 +1,85 @@
+
+namespace System.Text.RegularExpressions.Generator
+{
+    public static class DiagnosticDescriptors
```
I assume all these diagnostics shouldn't be suppressible. If that's the case, the descriptor should have custom tags containing [`WellKnownDiagnosticTags.NotConfigurable](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.wellknowndiagnostictags.notconfigurable?view=roslyn-dotnet-3.11.0)

Also was this intended to be public?

Youssef1313 commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs:
```diff
@@ -0,0 +1,68 @@
+namespace System.Text.RegularExpressions.Generator
+{
+    /// <summary>Generates C# source code to implement regular expressions.</summary>
+    [Generator]
```
Might be a good idea to explicitly supply `LanguageNames.CSharp` to the attribute. It's currently the same, but just in case roslyn changed that in future 😄

Youssef1313 commented on src/libraries/System.Text.RegularExpressions/tests/System.Text.RegularExpressions.Generators.Tests/RegexGeneratorParserTests.cs:
```diff
@@ -0,0 +1,507 @@
+            ", compile: true));
+        }
+
+        [Fact]
+        public async Task Valid_NestedVaryingTypes()
+        {
+            Assert.Empty(await RunGenerator(@"
+                using System.Text.RegularExpressions;
+                public partial class A
+                {
+                    public partial record class B
+                    {
+                        public partial record struct C
+                        {
+                            public partial record D
+                            {
+                                public partial struct E
+                                {
+                                    [RegexGenerator(""ab"")]
+                                    public static partial Regex Valid();
+                                }
+                            }
+                        }
+                    }
+                }
+            ", compile: true));
+        }
```
Consider testing a local function as well

stephentoub replied:
The methods need to be partial.  Partial locals aren't a thing.

Youssef1313 replied:
Yeah I meant that it produces a reasonable diagnostic, not that it compiles

stephentoub replied:
But it's not even valid syntax.

Youssef1313 replied:
@stephentoub Let me be more clear here, the attribute will be added on a non-partial local function, which is a valid syntax.

Currently, there is:

```csharp
                        // Scope to just methods
                        if (member is not MethodDeclarationSyntax methodSyntax)
                        {
                            continue;
                        }
```

which seems to silently ignore the attribute for that case. My expectation is to produce a diagnostic for that. Or if you want to silently skip it, then it's at least tested.

stephentoub replied:
Thanks, but I still don't understand why a generator should have to care about that at all.  Look how many errors you get if you even try to do this:
![image](https://user-images.githubusercontent.com/2642209/133617248-aafafa22-2866-4a4f-b951-a758d3699071.png)

stephentoub replied:
> the attribute will be added on a non-partial local function, which is a valid syntax.

You mean if the dev puts it on a non-partial local function?  I see; I can make sure the existing diagnostics fire correctly for that.

Youssef1313 replied:
@stephentoub I'm referring to a non-partial local function, which compiles fine:

```csharp
using System;
using System.Text.RegularExpressions;

[AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
public class MyAttribute : Attribute
{
}

public class C
{
    public void M()
    {
        Abc();

        [My]
        static Regex Abc() => null;        
    }
}
```

![image](https://user-images.githubusercontent.com/31348972/133617929-6b914748-0849-4849-a137-1e8a160a3799.png)

Do we want this to be just skipped as if the attribute isn't there? or do we want a diagnostic? whatever the desired behavior, it's good to be tested. I know this case isn't really important and both behaviors could be reasonable. But just for sake of completeness.

stephentoub replied:
> Do we want this to be just skipped as if the attribute isn't there?

The loop you cite won't be skipping local functions, as it's iterating over members, which won't ever contain local functions.  We'd need to scour the whole syntax tree looking for local functions, in which case it seems like we'd need to essentially change IsSyntaxTargetForGeneration/GetSemanticTargetForGeneration to essentially return every method or property and then have GetRegexTypesToEmit look for any local functions that are attributed.  Is that what you're proposing be done?  Note we don't have the ability to report a diagnostic in GetSemanticTargetForGeneration, nor do I believe we have the ability to cache things like type symbols from the compilation.

Szer commented:
> > Will there be any warning if that library code (supposed to be language agnostic?) would be called NOT from Roslyn?
> 
> I don't understand the question. `Regex` is usable from all .NET languages. If you try to use `RegexGenerator` with something other than C#, you'll get errors because no code will be generated. Can you clarify?

I was worrying that somebody could try to use BCL code from F# following your example
```fsharp
type Foo() =
    [<DefaultValue; RegexGenerator("abc", RegexOptions.IgnoreCase)>]
    val mutable regex: Regex
    
let foo = Foo()
foo.regex.Match("abc")
```
and it will fail only in runtime

It would be nice to have some kind of warning if these BCL attributes are only usable from specific compiler - Roslyn.

Youssef1313 commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexGeneratorAttribute.cs:
```diff
@@ -0,0 +1,44 @@
+namespace System.Text.RegularExpressions;
+
+/// <summary>Instructs the System.Text.RegularExpressions source generator to generate an implementation of the specified regular expression.</summary>
+[AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
```
Can it happen that a developer defines another version of the attribute that have `AllowMultiple = true` or different `AttributeTargets`, which can be returned by `GetTypeByMetadataName`?

I have no idea if this can happen, so just noting to consider.

stephentoub replied:
> Can it happen that a developer defines another version of the attribute that have AllowMultiple = true or different AttributeTargets, which can be returned by GetTypeByMetadataName?

In theory, sure. MultipleRegexGeneratorAttributes should flag multiple.  If someone creates their own and then uses it to attribute other things, that's on them; I don't think we should be significantly increasing the cost or complexity of the generator/analyzer for such things.

stephentoub committed:
Improve cachability of source generator

Changing the generator to not collect all regexes together means we don't need to reprocess/regenerate all regexes every time any one of them is changed.

eerhardt commented:
Regarding the Mono WASM failures

```C#
        private static MetadataReference[] CreateReferences()
        {
            string corelib = Assembly.GetAssembly(typeof(object))!.Location;
```

Assembly.Location is empty on WASM. You may need to log a bug and disable the tests on WASM, like we do for STJ and Logging generators:


https://github.com/dotnet/runtime/blob/f95449c1eda1f8cfd3e08af71ffaf8927f230036/src/libraries/tests.proj#L203-L209

eerhardt commented on src/libraries/System.Text.RegularExpressions/gen/Polyfills.cs:
```diff
@@ -0,0 +1,84 @@
+
+namespace System.Text.RegularExpressions
+{
+    internal class RegexReplacement
```
I think the rest of the code in this file makes sense, but this one kind of surprised me. It doesn't feel like the same category as the rest, since we are compiling the rest of the Regex code into this assembly. This one is more like a "stub" than a "polyfill" right?

stephentoub replied:
I guess.  I can rename the file to Stubs.cs if that addresses your concern.  This is "make the rest of the code compile and work correctly".cs.

eerhardt replied:
I think just a simple comment on this class would have alleviated my "surprise". The rest of them all made sense - new APIs that aren't available in netstandard2.0. This one just caused an "interrupt" in my mental model of this file.

eerhardt commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs:
```diff
@@ -0,0 +1,104 @@
+                // Pair each with the compilation
+                .Combine(context.CompilationProvider)
+
+                // Use a custom comparer that ignores the compilation so that it doesn't interface with the generators caching of results based on MethodDeclarationSyntax
```
From reading this comment, I'm not sure I understand what is happening here. (Even if I replace `interface` => `interfere` I still don't think I get why this is necessary.)

stephentoub replied:
Should have said interfere.  The idea is that we want to avoid regenerating code for a regex unless that regex has been changed.  Roslyn caches outputs based on SyntaxNodes, so as long as we can arrange things such that we produce output per SyntaxNode, we can make it so that we don't need to recompute for SyntaxNodes that haven't changed.  But changing any SyntaxNode will also change the Compilation, so we need to ignore the Compilation as part of the comparison operation Roslyn will do to determine whether things need to be recomputed.

eerhardt replied:
Is this purely a perf optimization?

stephentoub replied:
To avoid re-doing unnecessary work,  yes.

eerhardt commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs:
```diff
@@ -0,0 +1,104 @@
+    {
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            IncrementalValueProvider<ImmutableArray<object?>> typesAndCompilation =
```
`typesAndCompilation` seems out of date, right? At the end of this whole line, we don't have the "compilation" anymore.

stephentoub replied:
Yeah, that should have been renamed.

eerhardt commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs:
```diff
@@ -0,0 +1,286 @@
+                    break;
+                }
+
+                // If this isn't 
```
```suggestion
                // If this isn't a RegexGeneratorAttribute, skip it
```

eerhardt commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs:
```diff
@@ -0,0 +1,286 @@
+                return Diagnostic.Create(DiagnosticDescriptors.RegexMethodMustHaveValidSignature, methodSyntax.GetLocation());
+            }
+
+            if (typeDec.SyntaxTree.Options is CSharpParseOptions { LanguageVersion: < LanguageVersion.CSharp10 })
```
Is this because our code gen uses C# 10 features?

stephentoub replied:
I think technically I ended up using only C# 9, but I've not gone back to reconfirm.

eerhardt commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs:
```diff
@@ -0,0 +1,286 @@
+                return Diagnostic.Create(DiagnosticDescriptors.InvalidRegexArguments, methodSyntax.GetLocation(), e.Message);
+            }
+
+            // Determine the namespace the class is declared in, if any
```
This is one place in the Logging Generator I'd like changed. The Json generator does it better. Instead of interpreting the parents (and potentially having bugs when new features, like file scoped namespaces, are created), I think it is better to get the TypeSymbol of the containing type, and get its namespace that way. This is how the Json generator does it.

See https://github.com/dotnet/runtime/pull/57894#issuecomment-918247048

stephentoub commented:
> It would be nice to have some kind of warning if these BCL attributes are only usable from specific compiler - Roslyn.

@ericstj, where did we land on the approach of having generators inject the attribute they rely on (attributed as conditional if desired so it doesn't pollute the final assembly)? That avoids this kind of issue,  but we didn't take that approach with the other generators.  Should we reconsider moving forward?

eerhardt commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs:
```diff
@@ -0,0 +1,286 @@
+                return null;
+            }
+
+            TypeDeclarationSyntax typeDec = methodSyntax.Parent as TypeDeclarationSyntax;
+            if (typeDec is null)
```
Is it not possible to put a Regex attribute in a local method? If someone tries it, will they get a diagnostic?

stephentoub replied:
It is possible, and currently there won't be a diagnostic.  With how the Initialize logic was previously structured, this wasn't possible to do well.  Now that it's set up in a way that computes diagnostics earlier in the pipeline and then passes those into Execute, I can add an analysis pass that looks through the whole tree for local functions and sees whether any have the attribute applied.  It will add additional non-trivial cost, though, which is unfortunate when it's being done purely to catch erroneous usage (and you can't write partial locals).

eerhardt replied:
> (and you can't write partial locals)

ah right. Maybe we should just leave it as-is then.

eerhardt reviewed (commented):
I still have the Emitter code to go through, but wanted to log this first set of comments.

stephentoub commented:
> Regarding the Mono WASM failures

Yeah, I'd suppressed those failures for the main test project but forgot to do so for the generator test project.

danmoseley commented:
If code annotates two or more identical regexes, do we unify and write out only one implementation? I'm not sure how plausible that is, but it's possible. Eg., not everyone realizes that they are thread safe.

stephentoub commented:
> do we unify and write out only one implementation?

No. We could,  it's "just code", but currently I don't think it's worth the extra complexity for what should be a rare case the developer can address themselves. And I don't buy the thread-safety argument: a) the dev is writing a static method,  not a property, which they can call any number of times to get any number of instances... whether it returns the same or a new instance on each call is an implementation detail, and b) if that were really an issue,  they would likely need some way to source gen a new implementation per thread, which isn't a thing (outside of our tests).

stephentoub committed:
Use closest matching ctor from GetRegexAsync

To better test the appropriate ctor usage.

stephentoub committed:
Improve a few of the polyfills

stephentoub committed:
Address PR feedback

eerhardt commented:
>             writer.WriteLine($"{rm.Modifiers} Regex {rm.MethodName}() => {id}.Instance;");

With the Logging and Json generators, we fully qualify the generated type names. Do we want to here, as well?

ex. `global::System.Text.Json.Serialization.Metadata.JsonMetadataServices`.

---
Refers to: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:142 in 9684067. [](commit_id = 9684067d4c999e467fbc388411a0fad0d2c1ffde, deletion_comment = False)

eerhardt commented:
>             writer.Write($"    public static Regex Instance {{ get; }} = ");

(nit) Could be simplified by not using `$`:

`writer.Write("    public static Regex Instance { get; } = ");`

---
Refers to: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:148 in 9684067. [](commit_id = 9684067d4c999e467fbc388411a0fad0d2c1ffde, deletion_comment = False)

stephentoub commented:
> With the Logging and Json generators, we fully qualify type names in the generated code. Do we want to here, as well?

Which part are you suggesting be fully qualified?  The `{id}` part?  Is there a benefit to doing so here?

eerhardt commented:
> Which part are you suggesting be fully qualified?

The references to framework types. `Regex` => `global::System.Text.RegularExpressions.Regex`.

> Is there a benefit to doing so here?

The benefit is that if the user's project/namespace declares types that conflict with our type names, the code still compiles.

See also - https://github.com/dotnet/runtime/issues/52279

stephentoub commented:
> The references to framework types

Ah, I thought you were referring to the generated types.  Sure.

eerhardt reviewed (approved):
I can't truthfully say I understood all the inners of the Regex code, but this LGTM as an initial implementation.

stephentoub committed:
Fully qualify types and remove unnecessary $s

Also fixed one place where the IL we were generating wasn't as good as the reflection emit code.

stephentoub committed:
Suppress a couple more warnings

stephentoub commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -0,0 +1,3123 @@
+        {
+            "// <auto-generated/>",
+            "#nullable enable",
+            "#pragma warning disable CS0168 // Variable declared but never used",
+            "#pragma warning disable CS0164 // Unreferenced label",
```
> Is this "by design"? Or do we plan on changing this in the future to not suppress these warnings?

It's "by design" in that I'm aware of it :)  This happens with the existing reflection emit as well, there's just no warnings to suppress there.  We output labels to give us the opportunity to jump to things but then don't always emit code to jump to those locations.  We could clean it up, but it doesn't affect the generated IL quality, so I'm not particularly motivated to find ways to address it.  Making the code a bit cleaner would be the primary motivator.

stephentoub commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -0,0 +1,3123 @@
+            writer.WriteLine();
+            writer.WriteLine($"            protected override bool FindFirstChar()");
+            writer.WriteLine($"            {{");
+            writer.Indent += 4;
```
> Is it because we are already indented 3 "manually"?

Yeah.

For most of the code I rely on the indent from the writer, but here, it felt cleaner to just manually indent all the class boilerplate, to make it a bit easier to read in the generator.

stephentoub commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -0,0 +1,3123 @@
+
+            int AddGoto(int destpos)
+            {
+                if (forwardJumpsThroughSwitch![destpos] == -1)
```
That's left over from when it was nullable.  Thanks. Fixed.

stephentoub committed:
Fix stray !

stephentoub committed:
Fix TODO about unrolling multi comparisons

Also clean up generated code in a few places to make it more readable / concise.

Youssef1313 commented on src/libraries/System.Text.RegularExpressions/gen/DiagnosticDescriptors.cs:
```diff
@@ -0,0 +1,56 @@
+
+        public static DiagnosticDescriptor InvalidRegexArguments { get; } = new DiagnosticDescriptor(
+            id: "SYSLIB1042",
+            title: new LocalizableResourceString(nameof(SR.InvalidRegexArgumentsMessage), SR.ResourceManager, typeof(FxResources.System.Text.RegularExpressions.Generator.SR)),
```
This resource contain `{0}` placeholder. Only the `messageFormat` gets placeholders replaced. (I haven't validated other resources)

stephentoub replied:
Ok, thanks

Youssef1313 commented on src/libraries/System.Text.RegularExpressions/gen/Resources/Strings.resx:
```diff
@@ -0,0 +1,303 @@
+  <data name="RegexMethodMustHaveValidSignatureMessage" xml:space="preserve">
+    <value>Regex method must be static, partial, parameterless, non-generic, and return Regex.</value>
+  </data>
+  <data name="InvalidLangVersionMessage" xml:space="preserve">
+    <value>C# LangVersion of 10 or greater is required.</value>
+  </data>
```
There is an RSxxxx analyzer that should have flagged this.

https://github.com/dotnet/roslyn-analyzers/blob/a0522afd52b25ad211298e308fd5bbc7734865e0/src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx#L505-L507

https://github.com/dotnet/roslyn-analyzers/blob/a0522afd52b25ad211298e308fd5bbc7734865e0/src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx#L499-L501

Titles and messages shouldn't contain periods per the guidelines defined by that analyzer.

stephentoub replied:
This goes against the very fiber of my being :)  But ok.

stephentoub committed:
Update resources per PR feedback

Youssef1313 commented on src/libraries/System.Text.RegularExpressions/gen/DiagnosticDescriptors.cs:
```diff
@@ -0,0 +1,56 @@
+{
+    internal static class DiagnosticDescriptors
+    {
+        public static DiagnosticDescriptor InvalidRegexGeneratorAttribute { get; } = new DiagnosticDescriptor(
+            id: "SYSLIB1040",
+            title: new LocalizableResourceString(nameof(SR.InvalidRegexGeneratorAttributeTitle), SR.ResourceManager, typeof(FxResources.System.Text.RegularExpressions.Generator.SR)),
+            messageFormat: new LocalizableResourceString(nameof(SR.InvalidRegexGeneratorAttributeMessage), SR.ResourceManager, typeof(FxResources.System.Text.RegularExpressions.Generator.SR)),
+            category: "RegexGenerator",
+            DiagnosticSeverity.Error,
+            isEnabledByDefault: true,
+            customTags: WellKnownDiagnosticTags.NotConfigurable);
```
Will these be documented somewhere? Might be good to attach help links (in a follow-up PR, after they're documented)

stephentoub replied:
Eventually

Youssef1313 commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs:
```diff
@@ -0,0 +1,265 @@
+
+            return rc;
+
+            static bool IsAllowedKind(SyntaxKind kind) =>
+                kind == SyntaxKind.ClassDeclaration ||
+                kind == SyntaxKind.StructDeclaration ||
+                kind == SyntaxKind.RecordDeclaration ||
+                kind == SyntaxKind.RecordStructDeclaration;
```
Why not interfaces? This should be allowed unless I'm missing something

```csharp
public partial interface C
{
    [RegexGenerator]
    partial void M();
}

public partial interface C
{
    partial void M() { }
}
```

stephentoub replied:
Because... it didn't occur to me anyone would want to do this :)  But, I've added it.

Youssef1313 commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -0,0 +1,3187 @@
+            writer.WriteLine("{");
+            writer.Indent++;
+
+            // Generate a unique name to describe the regex instance.  This includes both
+            // the method name the user provided and a unique counter value, plus a simple
+            // non-randomized (for determinism) hash of the previous content to try to make
+            // the name that much harder to predict.
+            string generatedName = $"GeneratedRegex_{regexClass.Method.MethodName}_";
+            generatedName += ComputeStringHash(generatedName).ToString("X");
```
I don't get where the unique counter value comes from?

stephentoub replied:
The comment is stale.  I fixed it.

Youssef1313 commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs:
```diff
@@ -0,0 +1,265 @@
+                }
+
+                ImmutableArray<TypedConstant> items = attributeData.ConstructorArguments;
+                if (items.Length is > 0 and <= 3 && items[0].Value is string pattern)
```
What if named arguments is used and the arguments order is no longer matching parameter order?

stephentoub replied:
It works fine.  I've added a test to confirm.

stephentoub committed:
Add more tests, clean up parser code, and allow instance/interface methods

stephentoub committed:
Fix test suppression on mobile
